File: python/self_test.py
"""
Python S/MIME Self Test - validates sign, verify, encrypt, decrypt,
and sign+encrypt+decrypt+verify operations in isolation.
"""

import os
import sys
from smime_helper import (
    load_private_key, load_certificate,
    sign, verify_signature, encrypt, decrypt,
    sign_and_encrypt, decrypt_and_verify
)


def main():
    certs_dir = sys.argv[1] if len(sys.argv) > 1 else os.path.join(
        os.path.dirname(__file__), "..", "certs")

    print("=== Python S/MIME Self Test ===")

    pkey = load_private_key(os.path.join(certs_dir, "producer.pfx"), "changeit")
    pcert = load_certificate(os.path.join(certs_dir, "producer.pfx"), "changeit")
    ckey = load_private_key(os.path.join(certs_dir, "consumer.pfx"), "changeit")
    ccert = load_certificate(os.path.join(certs_dir, "consumer.pfx"), "changeit")
    ccert_pub = load_certificate(os.path.join(certs_dir, "consumer.crt"))

    payload = b'{"test": "hello"}'

    print("SIGN... ", end="")
    s, ct = sign(payload, "application/json", pkey, pcert)
    print(f"OK ({len(s)} bytes)")

    print("VERIFY... ", end="")
    v = verify_signature(s, ct, pcert)
    assert v == payload, "Verified payload does not match original"
    print(f"OK ({v})")

    print("ENCRYPT... ", end="")
    e, ct2 = encrypt(payload, "application/json", ccert_pub)
    print(f"OK ({len(e)} bytes)")

    print("DECRYPT... ", end="")
    d = decrypt(e, ct2, ckey, ccert)
    assert d == payload, "Decrypted payload does not match original"
    print(f"OK ({d})")

    print("SIGN+ENCRYPT... ", end="")
    se, ct3 = sign_and_encrypt(payload, "application/json", pkey, pcert, ccert_pub)
    print(f"OK ({len(se)} bytes)")

    print("DECRYPT+VERIFY... ", end="")
    dv = decrypt_and_verify(se, ct3, ckey, ccert, pcert)
    assert dv == payload, "Decrypt+Verify payload does not match original"
    print(f"OK ({dv})")

    print("ALL PYTHON S/MIME TESTS PASSED")


if __name__ == "__main__":
    main()

File: java/src/main/java/com/swim/smime/SelfTest.java
package com.swim.smime;

import java.nio.file.Path;
import java.security.PrivateKey;
import java.security.cert.X509Certificate;
import java.util.Arrays;

public class SelfTest {

    public static void main(String[] args) throws Exception {
        String certsDir = args.length > 0 ? args[0] : "../certs";

        System.out.println("=== Java S/MIME Self Test ===");

        PrivateKey signerKey = SmimeHelper.loadPrivateKey(
                Path.of(certsDir, "producer.pfx").toString(), "changeit");
        X509Certificate signerCert = SmimeHelper.loadCertificateFromPfx(
                Path.of(certsDir, "producer.pfx").toString(), "changeit");
        PrivateKey recipientKey = SmimeHelper.loadPrivateKey(
                Path.of(certsDir, "consumer.pfx").toString(), "changeit");
        X509Certificate recipientCert = SmimeHelper.loadCertificate(
                Path.of(certsDir, "consumer.crt").toString());

        byte[] payload = "{\"test\": \"hello\"}".getBytes("UTF-8");

        System.out.print("SIGN... ");
        SmimeHelper.SmimeResult signed = SmimeHelper.sign(payload, "application/json", signerKey, signerCert);
        System.out.println("OK (" + signed.data().length + " bytes)");

        System.out.print("VERIFY... ");
        byte[] verified = SmimeHelper.verifySignature(signed.data(), signed.contentType(), signerCert);
        assert Arrays.equals(verified, payload) : "Verified payload does not match original";
        System.out.println("OK (" + new String(verified, "UTF-8") + ")");

        System.out.print("ENCRYPT... ");
        SmimeHelper.SmimeResult encrypted = SmimeHelper.encrypt(payload, "application/json", recipientCert);
        System.out.println("OK (" + encrypted.data().length + " bytes)");

        System.out.print("DECRYPT... ");
        byte[] decrypted = SmimeHelper.decrypt(encrypted.data(), encrypted.contentType(), recipientKey);
        assert Arrays.equals(decrypted, payload) : "Decrypted payload does not match original";
        System.out.println("OK (" + new String(decrypted, "UTF-8") + ")");

        System.out.print("SIGN+ENCRYPT... ");
        SmimeHelper.SmimeResult se = SmimeHelper.signAndEncrypt(payload, "application/json",
                signerKey, signerCert, recipientCert);
        System.out.println("OK (" + se.data().length + " bytes)");

        System.out.print("DECRYPT+VERIFY... ");
        byte[] dv = SmimeHelper.decryptAndVerify(se.data(), se.contentType(), recipientKey, signerCert);
        assert Arrays.equals(dv, payload) : "Decrypt+Verify payload does not match original";
        System.out.println("OK (" + new String(dv, "UTF-8") + ")");

        System.out.println("ALL JAVA S/MIME TESTS PASSED");
    }
}
